04-higher-order/homework-04/hw-04.hs

```diff
@@ -0,0 +1,177 @@
+You should use the following data structure to represent binary
+trees. Note that each node stores an extra Integer representing the
+height at that node.
+data Tree a = Leaf
+| Node Integer (Tree a) a (Tree a)
+deriving (Show, Eq)
+For this exercise, write a function
+cis 194: homework 4 2
+foldTree :: [a] -> Tree a
+foldTree = ...
+which generates a balanced binary tree from a list of values using
+foldr.
+For example, one sample output might be the following, also visualized at right:
+D
+E
+A
+G
+H
+B
+F C
+I
+J
+foldTree "ABCDEFGHIJ" ==
+Node 3
+(Node 2
+(Node 0 Leaf ’F’ Leaf)
+’I’
+(Node 1 (Node 0 Leaf ’B’ Leaf) ’C’ Leaf))
+’J’
+(Node 2
+(Node 1 (Node 0 Leaf ’A’ Leaf) ’G’ Leaf)
+’H’
+(Node 1 (Node 0 Leaf ’D’ Leaf) ’E’ Leaf))
+Your solution might not place the nodes in the same exact order,
+but it should result in balanced trees, with each subtree having a
+correct computed height.
+-}
+--Exercise 2
+data Tree a = Leaf | Node Integer (Tree a) a (Tree a)
+    deriving (Show, Eq)
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert Leaf
+  where 
+    insert elem Leaf = Node 0 Leaf elem Leaf
+    insert elem (Node hgt left val right)
+        | getHgt left > getHgt right = Node hgt                                        left val (insert elem right)
+        | getHgt left < getHgt right = Node hgt                                        (insert elem left) val right
+        | otherwise                  = Node (1 + (getMaxHgt right $ insert elem left)) (insert elem left) val right
+          where
+            getHgt Leaf                    = -1
+            getHgt (Node hgt _ _ _)        = hgt
+            getMaxHgt firstTree secondTree = max (getHgt firstTree) (getHgt secondTree)
+
+
+{-
+Exercise 3: More folds!
+1. Implement a function
+xor :: [Bool] -> Bool
+which returns True if and only if there are an odd number of True
+values contained in the input list. It does not matter how many
+False values the input list contains. For example,
+xor [False, True, False] == True
+xor [False, True, False, False, True] == False
+Your solution must be implemented using a fold.
+2. Implement map as a fold. That is, complete the definition
+map’ :: (a -> b) -> [a] -> [b]
+map’ f = foldr ...
+in such a way that map’ behaves identically to the standard map
+function.
+cis 194: homework 4 3
+3. (Optional) Implement foldl using foldr. That is, complete the
+definition
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base xs = foldr ...
+in such a way that myFoldl behaves identically to the standard
+foldl function.
+Hint: Study how the application of foldr and foldl work out:
+foldr f z [x1, x2, ..., xn] == x1 ‘f‘ (x2 ‘f‘ ... (xn ‘f‘ z)...)
+foldl f z [x1, x2, ..., xn] == (...((z ‘f‘ x1) ‘f‘ x2) ‘f‘...) ‘f‘ xn
+-}
+--Exercise 3
+xor :: [Bool] -> Bool
+xor = foldr (/=) False
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x xs -> f x : xs) []
+
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f initial xs = foldr (flip f) initial (reverse xs)
```

Comment:
This implementation works fine, but there's a way to do it without relying on any existing list functions (apart from foldr of course :) )
This is rather hard, but can be done if you realize you need to accumulate a function and in the end you apply it to the initial element. Only try this is you want extra challenge. If you give you google for "implementing foldl in terms of foldl".

Reply:
@jhrcek Thanks for the feedback, now I know at which direciton to look into. I wasn't very satisfied with my answer anyways 😂